awx/ui/client/features/projects/projectsList.controller.js

```diff
@@ -0,0 +1,439 @@
+/** ***********************************************
+ * Copyright (c) 2018 Ansible, Inc.
+ *
+ * All Rights Reserved
+ ************************************************ */
+
+const mapChoices = choices => Object.assign(...choices.map(([k, v]) => ({ [k]: v.toUpperCase() })));
+
+function projectsListController (
+    $filter, $scope, $rootScope, $state, $log, Dataset, Alert, Rest,
+    ProcessErrors, resolvedModels, strings, Wait, ngToast,
+    Prompt, GetBasePath, qs, ProjectUpdate,
+) {
+    const vm = this || {};
+    const [ProjectModel] = resolvedModels;
+    $scope.canAdd = ProjectModel.options('actions.POST');
+
+    vm.strings = strings;
+    vm.scm_choices = ProjectModel.options('actions.GET.scm_type.choices');
+    vm.projectTypes = mapChoices(vm.scm_choices);
+
+    // smart-search
+    vm.list = {
+        iterator: 'project',
+        name: 'projects',
+        basePath: 'projects',
+    };
+    vm.dataset = Dataset.data;
+    vm.projects = Dataset.data.results;
+    // build tooltips
+    _.forEach(vm.projects, buildTooltips);
+    $rootScope.flashMessage = null;
+
+    // when a project is added/deleted, rebuild tooltips
+    $scope.$watchCollection('vm.projects', () => {
+        _.forEach(vm.projects, buildTooltips);
+    });
+    // show active item in the list
+    $scope.$watch('$state.params', () => {
+        const projectId = _.get($state.params, 'project_id');
+        if ((projectId)) {
+            vm.activeId = parseInt($state.params.project_id, 10);
+        } else {
+            vm.activeId = '';
+        }
+    }, true);
+
+    $scope.$on('ws-jobs', (e, data) => {
+        $log.debug(data);
+        if (vm.projects) {
+            // Assuming we have a list of projects available
+            const project = vm.projects.find((p) => p.id === data.project_id);
+            if (project) {
+                // And we found the affected project
+                $log.debug(`Received event for project: ${project.name}`);
+                $log.debug(`Status changed to: ${data.status}`);
+                if (data.status === 'successful' || data.status === 'failed' || data.status === 'canceled') {
+                    reloadList();
+                } else {
+                    project.scm_update_tooltip = vm.strings.get('update.UPDATE_RUNNING');
+                }
+                project.status = data.status;
+                buildTooltips(project);
+            }
+        }
+    });
+
+    if ($scope.removeGoTojobResults) {
```

Comment:
This seems a bit weird to me and looks like it's trying to handle instantiating an event handler on a `broadcast` or `emit` event. It would be more linear to write something like:

```
$scope.$on(`GoTojobResults`, handleGoToJobResults(e, data));
function handleGoToJobResults(event, res) = {
// handle event
}